### PR TITLE
PersistentCollection now extends AbstractLazyCollection.

### PR DESCRIPTION
**Pull request by guilhermeblanco**
Requires https://github.com/doctrine/collections/pull/52

### CODE REVIEW

**Comment by guilhermeblanco** [comment]
WTH at failed test? I'll have to investigate...

**Comment by guilhermeblanco** [comment]
WTH at failed test? I'll have to investigate...

**Comment by guilhermeblanco** [comment]
@Ocramius can you try to figure out what is happening here? 
Couldn't understand how it affected the hash cleanup. I also need to finish next PR before we enter beta stage

**Comment by Ocramius** [comment]
> @Ocramius can you try to figure out what is happening here?

Can check later today, I think.

**Comment by Ocramius** [comment]
This test fails on `Doctrine\Tests\ORM\Functional\IdentityMapTest#testReusedSplObjectHashDoesNotConfuseUnitOfWork()`.

The test is very fragile, and makes assumptions about the number of internally instantiated/referenced objects, so I think it's better to kill those tests completely (as they are currently failing on PHP7 as well)

**Comment by guilhermeblanco** [comment]
@Ocramius will remove them... =)

**Comment by bakura10** [comment]
I like this and I've love this to be merged for Doctrine 2.5. It would be cleaner to be able to typehint against this class rather than PersistentCollection when we need to do a check if association is initialized or not.

**Comment by Ocramius** [comment]
@bakura10 yeah, but we still have no clue why tests are failing, therefore it is `Delayed`

**Comment by guilhermeblanco** [comment]
Spent a lot of time looking at possible root cause of this issue.
As soon as I start to use `collection` from `AbtractLazyCollection` instead of declared `coll` in `PersistentCollection`, the IdentityMapTest starts to fail.
At this point, I assume there's an internal leak inside PHP when reusing parent properties. I'll remove the test since it's very fragile and consider this PR ready

**Comment by guilhermeblanco** [comment]
@Ocramius I consider this PR done and ready for review now

File: lib/Doctrine/ORM/PersistentCollection.php [hunk 1]
```diff
@@ -121,9 +107,10 @@
      */
     public function __construct(EntityManagerInterface $em, $class, $coll)
     {
-        $this->coll      = $coll;
-        $this->em        = $em;
-        $this->typeClass = $class;
+        $this->collection  = $coll;
+        $this->em          = $em;
+        $this->typeClass   = $class;
+        $this->initialized = true;
```
Comment (Ocramius) [inline lib/Doctrine/ORM/PersistentCollection.php]: Seems wrong to me...

Comment (guilhermeblanco) [inline lib/Doctrine/ORM/PersistentCollection.php]: @Ocramius keeping same code as it was before. Look at deleted line 106 at same file. =)

Comment (Ocramius) [inline lib/Doctrine/ORM/PersistentCollection.php]: @guilhermeblanco weird stuff D:

**Comment by Ocramius** [comment]
:+1:

**Comment by guilhermeblanco** [comment]
We should keep this PR somehow in our mind because of the memory leak that seems to happen here.

**Comment by karlrixon** [comment]
This PR is breaking things for me - it many places in PersistentCollection it tries to access the initialized property which is a private member of the parent class. The end result is that calls to isInitialized() always return false for me, even though the collection is initialized...

**Comment by guilhermeblanco** [comment]
@karlrixon you have to upgrade doctrine/collections too

**Comment by karlrixon** [comment]
@guilhermeblanco, thanks, updating to the master release of doctrine/collections does indeed fix things.

**Pull request merged**